schema/field.go

```diff
@@ -93,25 +117,12 @@ type FieldSerializer interface {
 	Serialize(value interface{}) (interface{}, error)
 }
 
-// Compile implements the ReferenceCompiler interface and recursively compile sub schemas
-// and validators when they implement Compiler interface.
-func (f Field) Compile(rc ReferenceChecker) error {
-	// TODO check field name format (alpha num + _ and -).
-	if f.Schema != nil {
-		// Recursively compile sub schema if any.
-		if err := f.Schema.Compile(rc); err != nil {
-			return fmt.Errorf(".%v", err)
-		}
-	} else if f.Validator != nil {
-		// Compile validator if it implements the ReferenceCompiler or Compiler interface.
-		if c, ok := f.Validator.(Compiler); ok {
-			if err := c.Compile(rc); err != nil {
-				return fmt.Errorf(": %v", err)
-			}
-		}
-		if reflect.ValueOf(f.Validator).Kind() != reflect.Ptr {
-			return errors.New(": not a schema.Validator pointer")
-		}
-	}
-	return nil
+// FieldGetter is used to describe a FieldValidator (or Schema) which allow JSON
+// object values.
+type FieldGetter interface {
+	// GetField returns the validator for the field if the given field name is
+	// present in the schema.
+	//
+	// You may reference sub field using dotted notation field.subfield.
+	GetField(name string) *Field
```

Comment:
`QueryField` could be a better name. At least when it comes to e.g. `Array` and `AnyOf`, we are interested in a Field that can be used to validate query values, not necessarily a field that can validate values.

- https://docs.mongodb.com/manual/tutorial/query-arrays/

Reply:
Requires some rewrite of the whole logic to support better array querying, so we could wait with the rename this until we do the array part, if we need to slightly redo the logic anyway...